### PR TITLE
Return the `default_root_dir` as the `log_dir` when the logger is a `LoggerCollection`

### LINKED ISSUE #8157: LoggerCollection breaks PyTorch profiler
## 🐛 Bug

If the Trainer's `profiler` parameter is set to `"pytorch"` and the `Trainer`'s logger is an instance of `LoggerCollection`, the profiler fails to write to a local file (with a warning).

The path for said file is derived from [this property](https://github.com/PyTorchLightning/pytorch-lightning/blob/28afc7a10d9f9c1160935fb5c81a1a8c0492b392/pytorch_lightning/trainer/properties.py#L223) of the Trainer, which in turn derives from the `save_dir` of the Trainer's logger whenever the logger isn't a stock TensorBoard logger. The `save_dir` property of a LoggerCollection, which is what you get when the trainer is provided with more than one logger, [is set to None by default](https://github.com/PyTorchLightning/pytorch-lightning/blob/28afc7a10d9f9c1160935fb5c81a1a8c0492b392/pytorch_lightning/loggers/base.py#L416), apparently deliberately. I'm not sure what the rationale behind that decision was, so I'm hesitant to make a PR myself, but it seems logical to me that `LoggerCollection` should have a `log_path` property defined analogously to [that of a `TensorBoardLogger`](https://github.com/PyTorchLightning/pytorch-lightning/blob/28afc7a10d9f9c1160935fb5c81a1a8c0492b392/pytorch_lightning/loggers/tensorboard.py#L119), and that the property of the `Trainer` shouldn't fall back to the `save_dir` property automatically when its logger isn't an instance of `TensorBoardLogger`.

## Reproduction

[Here's a BoringModel Colab link](https://colab.research.google.com/drive/1qt2wedM_CahAphWwJTjTk3W-3vlpLJjO?usp=sharing). The following warning is generated during training when the `Trainer`'s logger is wrapped in a list:

```
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/profiler/pytorch.py:441: UserWarning: The PyTorchProfiler failed to export trace as `dirpath` is None
  rank_zero_warn("The PyTorchProfiler failed to export trace as `dirpath` is None")
```

### Expected behavior

The PyTorch profiler should always have access to a default logging path, even when the `Trainer`'s logger is a `LoggerCollection`. Ideally, it would follow the same "version" convention obeyed when the logger is a `TensorBoardLogger`.

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.3.7post0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

### PR DESCRIPTION
**Pull request by gahdritz**
## What does this PR do?

Fixes a bug whereby the PyTorch profiler would fail to write to a local file when the `Trainer`'s logger is a `LoggerCollection`. 

Fixes #8157

The way I see it, there are three ways to go about this:

1. Change the `save_dir` property of the LoggerCollection to the `default_root_dir` of the `Trainer`
2. What I did here
3. Some more complex fix, arguably the most natural, where the profiler traces are written to the `save_dir` of each constituent logger.

The first would probably break code that relies on the property being `None` to set defaults. The third, since the Lightning wrapper around the PyTorch profiler relies on PyTorch functions that expect a single path, is probably prohibitively complex.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/profiler/test_profiler.py [hunk 1]
```diff
@@ -431,6 +433,29 @@ def test_pytorch_profiler_nested(tmpdir):
     assert events_name == expected, (events_name, torch.__version__, platform.system())
 
 
+def test_pytorch_profiler_loggercollection(tmpdir):
```
Comment (awaelchli) [inline tests/profiler/test_profiler.py:436]: Hey! Thanks for the PR. 
This test seems useful, however the fix itself is quite independent of the profiling. I would like to suggest adding a small test to `tests/trainer/properties/log_dir/` for the logger collection save path.

Comment (gahdritz) [reply tests/profiler/test_profiler.py:436]: Done!

File: tests/trainer/properties/log_dir.py [hunk 2]
```diff
@@ -140,3 +140,23 @@ def test_logdir_custom_logger(tmpdir):
...
+    expected = tmpdir
+    model = TestModel(expected)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=2,
+        callbacks=[ModelCheckpoint(dirpath=tmpdir)],
+        logger=[TensorBoardLogger(save_dir=tmpdir, name='custom_logs')]
+    )
+
```
Comment (tchaton) [inline tests/trainer/properties/log_dir.py:159]: Can you check the trainer.logger is a LoggerCollection ? There is only 1 logger provided there.

File: tests/trainer/properties/log_dir.py [hunk 3]
```diff
@@ -140,3 +140,23 @@ def test_logdir_custom_logger(tmpdir):
     assert trainer.log_dir == expected
     trainer.fit(model)
     assert trainer.log_dir == expected
+
+
+def test_logdir_loggercollection(tmpdir):
```
Comment (tchaton) [inline tests/trainer/properties/log_dir.py:145]:
```suggestion
def test_logdir_logger_collection(tmpdir):
```

File: tests/profiler/test_profiler.py [hunk 4]
```diff
@@ -431,6 +432,36 @@ def test_pytorch_profiler_nested(tmpdir):
...
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        profiler="pytorch",
+        logger=logger,
+        limit_train_batches=5,  # it takes a while for the warning to appear
+        max_epochs=1,
+    )
+
+    trainer.fit(model)
+
```
Comment (tchaton) [inline tests/profiler/test_profiler.py:461]: Same here. Add a check for `assert isinstance(trainer.logger, LoggerCollection)`

Comment (gahdritz) [reply tests/profiler/test_profiler.py:463]: Done!

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/properties.py [hunk 5]
```diff
@@ -223,8 +224,13 @@ def model(self, model: torch.nn.Module) -> None:
     def log_dir(self) -> Optional[str]:
         if self.logger is None:
             dirpath = self.default_root_dir
+        elif isinstance(self.logger, TensorBoardLogger):
+            dirpath = self.logger.log_dir
```
Comment (Borda) [inline pytorch_lightning/trainer/properties.py:228]: this is inconsistent, then let's add `save_dir` property to `TensorBoardLogger`

File: tests/trainer/properties/log_dir.py [hunk 6]
```diff
@@ -140,3 +140,24 @@ def test_logdir_custom_logger(tmpdir):
...
+    LoggerCollection
+    """
+    expected = tmpdir
+    model = TestModel(expected)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=2,
+        callbacks=[ModelCheckpoint(dirpath=tmpdir)],
+        logger=[TensorBoardLogger(save_dir=tmpdir, name='custom_logs')]
```
Comment (Borda) [inline tests/trainer/properties/log_dir.py:157]: this shall work for any logger, right?

File: pytorch_lightning/trainer/properties.py [hunk 7]
```diff
@@ -223,8 +224,13 @@ def model(self, model: torch.nn.Module) -> None:
...
         else:
-            dirpath = getattr(self.logger, 'log_dir' if isinstance(self.logger, TensorBoardLogger) else 'save_dir')
+            dirpath = self.logger.save_dir
+
+        if isinstance(self.logger, LoggerCollection) and dirpath is None:
+            dirpath = self.default_root_dir
```
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:233]: Why not move this if to the `if-elif` chain above?

File: pytorch_lightning/trainer/properties.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/properties.py:228]: If i'm not missing something, the current changes keep the same functionality.

Adding `save_dir` would be for a separate PR IMO

File: tests/trainer/properties/log_dir.py [hunk 8]
```diff
@@ -140,3 +140,24 @@ def test_logdir_custom_logger(tmpdir):
     assert trainer.log_dir == expected
     trainer.fit(model)
     assert trainer.log_dir == expected
+
+
+def test_logdir_logger_collection(tmpdir):
+    """
+    Tests that the logdir equals the default_root_dir when the logger is a
+    LoggerCollection
+    """
```
Comment (carmocca) [inline tests/trainer/properties/log_dir.py:149]: Fits in one line

```suggestion
    """Tests that the logdir equals the default_root_dir when the logger is a LoggerCollection"""
```

File: tests/trainer/properties/log_dir.py [hunk 9]
```diff
@@ -140,3 +140,24 @@ def test_logdir_custom_logger(tmpdir):
...
+    expected = tmpdir
+    model = TestModel(expected)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=2,
+        callbacks=[ModelCheckpoint(dirpath=tmpdir)],
```
Comment (carmocca) [inline tests/trainer/properties/log_dir.py:156]: Is a `ModelCheckpoint` callback necessary for the test?

File: tests/profiler/test_profiler.py [hunk 10]
```diff
@@ -431,6 +433,38 @@ def test_pytorch_profiler_nested(tmpdir):
...
 
+def test_pytorch_profiler_logger_collection(tmpdir):
+    """
+    Tests whether the PyTorch profiler is able to write its trace locally when
+    the Trainer's logger is an instance of LoggerCollection. See issue #8157.
+    """
+
+    def look_for_trace(trace_dir):
+        """ Determines if a directory contains a PyTorch trace """
+        return any(("trace.json" in filename for filename in os.listdir(trace_dir)))
```
Comment (carmocca) [inline tests/profiler/test_profiler.py:444]: Extra parenthesis
```suggestion
        return any("trace.json" in filename for filename in os.listdir(trace_dir))
```

File: tests/profiler/test_profiler.py [hunk 11]
```diff
@@ -431,6 +433,38 @@ def test_pytorch_profiler_nested(tmpdir):
...
+
+    model = BoringModel()
+
+    # Wrap the logger in a list so it becomes a LoggerCollection
+    logger = [TensorBoardLogger(save_dir=tmpdir)]
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        profiler="pytorch",
+        logger=logger,
+        limit_train_batches=5,  # it takes a while for the warning to appear
```
Comment (carmocca) [inline tests/profiler/test_profiler.py:457]: What do you mean by this comment? Which warning exactly?

Comment (gahdritz) [reply tests/profiler/test_profiler.py:457]: The test previously looked for a warning that would appear when you used a LoggerCollection here. I'll get rid of the comment.

File: tests/trainer/properties/log_dir.py (hunk 9 above)
Comment (gahdritz) [reply tests/trainer/properties/log_dir.py]: Nope---I've removed it.

File: pytorch_lightning/trainer/properties.py (hunk 7 above)
Comment (gahdritz) [reply pytorch_lightning/trainer/properties.py]: Its condition was a little more specific than the others, so I thought it would make more sense for it to be separate. On second thought, though, letting users override the default behavior by manually setting the `save_dir` of their LoggerCollection probably isn't very clean. I've removed the second condition and merge the blocks.

File: tests/trainer/properties/log_dir.py (hunk 6 above)
Comment (gahdritz) [reply tests/trainer/properties/log_dir.py]: Yes---it doesn't necessarily have to be a TensorBoardLogger.

File: tests/trainer/properties/log_dir.py [hunk 12]
```diff
@@ -140,3 +140,18 @@ def test_logdir_custom_logger(tmpdir):
...
     assert trainer.log_dir == expected
+
+
+def test_logdir_logger_collection(tmpdir):
+    """Tests that the logdir equals the default_root_dir when the logger is a LoggerCollection"""
+    expected = tmpdir
+    model = TestModel(expected)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir, max_steps=2, logger=[TensorBoardLogger(save_dir=tmpdir, name='custom_logs')]
```
Comment (awaelchli) [inline tests/trainer/properties/log_dir.py:151]: if you are testing that the expected log dir should be the default_root_dir, you should set the same value for save_dir in the tensorboard logger. choose a different path so the test can never give a false positive. 
hope it's clear what I mean :)

**Review by Borda — APPROVED** [review]

File: tests/trainer/properties/log_dir.py (hunk 12 above)
Comment (kaushikb11) [reply tests/trainer/properties/log_dir.py:151]: @gahdritz Could you address @awaelchli's comment?

Comment (carmocca) [reply tests/trainer/properties/log_dir.py]: @awaelchli can you confirm https://github.com/PyTorchLightning/pytorch-lightning/pull/8187/commits/6941d930a4edfd94d960132bc487025e025aabec is what you meant?

Comment (awaelchli) [reply tests/trainer/properties/log_dir.py]: yes, thanks
that's what I meant

**Review by awaelchli — APPROVED** [review]
